File: algorithms/python/cousinsinBinaryTree/cousinsinBinaryTree.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/cousins-in-binary-tree/
# Author: Miao Zhang
# Date:   2021-03-31

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        self.px = None
        self.dx = -1
        self.py = None
        self.dy = -1
        
        def preorder(root, parent, d):
            if not root: return
            if root.val == x:
                self.px, self.dx = parent, d
            if root.val == y:
                self.py, self.dy = parent, d
            preorder(root.left, root, d + 1)
            preorder(root.right, root, d + 1)
        
        preorder(root, None, 0)
        return self.dx == self.dy and self.px != self.py

File: algorithms/cpp/cousinsinBinaryTree/cousinsinBinaryTree.cpp
// Source: https://leetcode.com/problems/cousins-in-binary-tree/
// Author: Miao Zhang
// Date:   2021-03-31

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        preorder(root, nullptr, 0, x, y);
        return px_ != py_ && dx_ == dy_;
    }

private:
    TreeNode* px_;
    TreeNode* py_;
    int dx_;
    int dy_;
    
    void preorder(TreeNode* root, TreeNode* parent, int d, int x, int y) {
        if (!root) return;
        if (root->val == x) {
            px_ = parent;
            dx_ = d;
        }
        if (root->val == y) {
            py_ = parent;
            dy_ = d;
        }
        preorder(root->left, root, d + 1, x, y);
        preorder(root->right, root, d + 1, x, y);
    }
    
};
